File: day13/day13.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        
        int n = nums.size();
        vector<int>vec(n+1,0);
        for(int i = 0;i < n;i++) {
            vec[i+1] = vec[i] + ( nums[i] == 1 ? 1 : -1);
        }
        
        unordered_map<int,int>hash;
        int ans = 0;
        for(int i=0;i<=n;i++) {
            if(hash.find(vec[i]) == hash.end()) {
                hash[vec[i]] = i;
            }
            else {
                int index = hash[vec[i]];
                ans = max(ans,i-index);
            }
        }
        return ans;
    }
};

File: day13/Day13.java
class Day13 {
    public int findMaxLength(int[] nums) {
        
        int n = nums.length;
        int[] arr = new int[n+1];
        
        for(int i = 0;i<n;i++) {
            arr[i+1] = arr[i] + (nums[i] == 0 ? -1:1);
        }
        Map<Integer,Integer> myMap = new HashMap<Integer,Integer>();
        int ans = 0;
        for(int i = 0;i <= n;i++) {
            
            if(myMap.containsKey(arr[i] ) == true) {
                int index = myMap.get(arr[i]);
                ans = Math.max(ans,i-index);
            }
            else {
                myMap.put(arr[i],i);
            }
        }
        return ans;
    }
}